### PR TITLE
Simplify `CapabilityRegistry`

### PR DESCRIPTION
**Pull request by ahoppen**
Improve `CapabilityRegistry` and surrounding code in a couple of ways
- Make `CapabilityRegistry` an actor
- Refactor `CapabilityRegistry` to duplicate less code
- Ensure that `SourceKitServer` reports all capabilities of `SwiftLanguageServer` if the client doesn’t support dynamic capability registration. We were missing this for semantic tokens and inlay hits
- Add a few doc comments

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SourceKitLSP/CapabilityRegistry.swift [hunk 1]
```diff
@@ -12,14 +12,17 @@
...
-public typealias ClientRegistrationHandler = (CapabilityRegistration) -> Void
+import SKSupport
 
 /// A class which tracks the client's capabilities as well as our dynamic
 /// capability registrations in order to avoid registering conflicting
 /// capabilities.
-public final class CapabilityRegistry {
+public final actor CapabilityRegistry {
+  /// The client's capabilities as they were reported when sourcekit-lsp was launched.
+  public let clientCapabilities: ClientCapabilities
```
Comment (bnbarham) [inline Sources/SourceKitLSP/CapabilityRegistry.swift:22]: Maybe `staticClientCapabilities`? Or `initialClientCapabilities` or... something.

File: Sources/SourceKitLSP/CapabilityRegistry.swift [hunk 2]
```diff
@@ -75,48 +74,128 @@ public final class CapabilityRegistry {
...
-  /// languages.
-  public func registerCompletionIfNeeded(
-    options: CompletionOptions,
+  /// Since LSP 3.17.0, diagnostics can be reported through pull-based requests in addition to the existing push-based
+  /// publish notifications.
+  ///
+  /// The `DiagnosticOptions` were added at the same time as the pull diagnostics request and allow specification of
+  /// options for the pull diagnostics request. If the client doesn't reject this dynamic capability registration,
+  /// it supports the pull diagnostics request.
+  public func clientSupportsPullDiagnostics(for language: Language) -> Bool {
```
Comment (bnbarham) [inline Sources/SourceKitLSP/CapabilityRegistry.swift:89]: Can a client support pull based diagnostics but *not* dynamic registration? Or is pull based on available through dynamic registration?

File: Sources/SourceKitLSP/CapabilityRegistry.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/CapabilityRegistry.swift:22]: There is no dynamic registration of client capabilities, so it doesn’t make sense to differentiate them.

**Pull request merged**